app/models/katello/concerns/pulp_database_unit.rb

```diff
@@ -120,34 +125,38 @@ def import_for_repository(repository)
               service.update_model(model)
             end
             pulp_ids << pulp_id
+            pulp_id_href_map.merge!(pulp_id => backend_identifier) if backend_identifier
           end
         end
-        sync_repository_associations(repository, :pulp_ids => pulp_ids) if self.many_repository_associations
+        sync_repository_associations(repository, :pulp_ids => pulp_ids, :pulp_id_href_map => pulp_id_href_map) if self.many_repository_associations
       end
 
       def sync_repository_associations(repository, options = {})
         additive = options.fetch(:additive, false)
-        pulp_ids = options.fetch(:pulp_ids, nil)
-        associated_ids = with_pulp_id(pulp_ids).pluck(:id) if pulp_ids
-
-        table_name = self.repository_association_class.table_name
-        attribute_name = unit_id_field
-
+        pulp_ids = options.dig(:pulp_ids)
+        pulp_id_href_map = options.dig(:pulp_id_href_map)
+        id_map_for_repository = {}
+        if pulp_ids
```

Comment:
when would this be false?

Reply:
I believe when a content unit gets deleted in a repo version upon sync..

Reply:
ACK. The way this code is (and was) written is that it assumes that `associated_ids` will always be defined on this line: 

```  
new_ids = associated_ids - existing_ids
```

This is making me think about how to clean up this code because it is a little hard to follow due to the organization and some of the data manipulation going on. I think the time is now for a refactor of the method. Let me know what you think about this. Do you agree with it? Does it still work?

```
     def sync_repository_associations(repository, options = {})                                                                                                                                                                                                                
        additive = options.fetch(:additive, false)                                                                                                                                                                                                                              
        pulp_ids = options.fetch(:pulp_ids, []) # seems like we always give pulp_ids, ie it is required                                                                                                                                                                         
                                                                                                                                                                                                                                                                                
        existing_ids = self.repository_association_class.uncached do                                                                                                                                                                                                            
          self.repository_association_class.where(:repository_id => repository).pluck(unit_id_field)                                                                                                                                                                            
        end                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                
        ids_for_repository = with_pulp_id(pulp_ids).pluck(:id, :pulp_id)                                                                                                                                                                                                        
        associated_ids = ids_for_repository.map(&:first)                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                
        unless additive                                                                                                                                                                                                                                                         
          delete_ids = existing_ids - associated_ids                                                                                                                                                                                                                            
          ActiveRecord::Base.connection.execute(                                                                                                                                                                                                                                
            "DELETE FROM #{self.repository_association_class.table_name}                                                                                                                                                                                                        
            WHERE repository_id=#{repository.id}                                                                                                                                                                                                                                
            AND #{unit_id_field} IN (#{delete_ids.join(', ')})"                                                                                                                                                                                                                 
          ) if delete_ids.any?                                                                                                                                                                                                                                                  
        end                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                
        new_ids = associated_ids - existing_ids                                                                                                                                                                                                                                 
        unless new_ids.empty?                                                                                                                                                                                                                                                   
          id_map_for_repository = {}                                                                                                                                                                                                                                            
          pulp_id_href_map = options.fetch(:pulp_id_href_map, {})                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                
          if pulp_id_href_map.any?                                                                                                                                                                                                                                              
            ids_for_repository.map do |pair|                                                                                                                                                                                                                                    
              # pair[0] is the unit id                                                                                                                                                                                                                                          
              # pair[1] is the pulp_id                                                                                                                                                                                                                                          
              id_map_for_repository[pair[0]] = pulp_id_href_map[pair[1]]                                                                                                                                                                                                        
            end                                                                                                                                                                                                                                                                 
          end                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                
          self.repository_association_class.import(db_columns_sync, db_values(new_ids, id_map_for_repository, repository), validate: false)                                                                                                                                     
        end                                                                                                                                                                                                                                                                     
      end
```